src/profile/jwks.ts

```diff
@@ -0,0 +1,185 @@
+  profileDocument: SolidDataset,
+  webId: WebId
+): UrlString | null {
+  return getUrl(
+    getProfileFromProfileDoc(profileDocument, webId),
+    security.publicKey
+  );
+}
+
+const isJwks = (jwksDocument: Jwks | unknown): jwksDocument is Jwks => {
+  return typeof (jwksDocument as Jwks).keys !== "undefined";
+};
+
+/**
+ * Fetch a JWKS at a given IRI, and add the given JWK to the obtained key set.
+ *
+ * @param jwk The JWK to add to the set.
+ * @param jwksIri The IRI where the key set should be looked up.
+ * @param options @param options Optional parameter `options.fetch`: An alternative `fetch` function to make the HTTP request, compatible with the browser-native [fetch API](https://developer.mozilla.org/docs/Web/API/WindowOrWorkerGlobalScope/fetch#parameters).
+ * @returns Promise resolving to a JWKS where the given key has been added.
+ */
+export async function addJwkToJwks(
+  jwk: Jwk,
+  jwksIri: IriString,
+  options: Partial<
+    typeof internal_defaultFetchOptions
+  > = internal_defaultFetchOptions
+): Promise<Jwks> {
+  const config = {
+    ...internal_defaultFetchOptions,
+    ...options,
+  };
+  const jwksResponse = await config.fetch(jwksIri);
+  if (!jwksResponse.ok) {
+    throw new Error(
+      `Fetching [${jwksIri}] returned an error: ${jwksResponse.status} ${jwksResponse.statusText}`
+    );
+  }
+  try {
+    const jwksDocument = await jwksResponse.json();
+    if (!isJwks(jwksDocument)) {
+      throw new Error(
+        `${jwksIri} did not dereference to a valid JWKS: ${JSON.stringify(
+          jwksDocument
+        )}`
+      );
+    }
+    return {
+      keys: [...jwksDocument.keys, jwk],
+    };
+  } catch (e) {
+    throw new Error(`Parsing the document at [${jwksIri}] failed: ${e}`);
+  }
+}
+
+/**
+ * Adds a public key to the JWKS listed in the profile associated to the given WebID.
+ * After having looked up the profile document thanks to the WebID, the associated
+ * JWKS, and having added the given key to the JWKS, this function overwrites the
+ * previous JWKS so that the new version is saved. This assumes the JWKS is hosted
+ * at a read-write IRI, such as a Solid pod.
+ *
+ * @param publicKey The public key value to set.
+ * @param webId The WebID of the profile to add the public key.
+ * @param options Optional parameter `options.fetch`: An alternative `fetch` function to make the HTTP request, compatible with the browser-native [fetch API](https://developer.mozilla.org/docs/Web/API/WindowOrWorkerGlobalScope/fetch#parameters).
+ */
+export async function addPublicKeyToProfileJwks(
+  publicKey: Jwk,
+  webId: WebId,
+  options: Partial<
+    typeof internal_defaultFetchOptions
+  > = internal_defaultFetchOptions
+): Promise<Blob & WithResourceInfo> {
+  const profileDataset = await getSolidDataset(webId, {
+    fetch: options.fetch,
+  });
+  if (profileDataset === null) {
+    throw new Error(
+      `The profile document associated to [${webId}] could not be found.`
```

Comment:
I'd mentioned that the IRI we're talking about is specifically a 'WebID' IRI:
```suggestion
      `The profile document associated with WebID [${webId}] could not be found.`
```
Also, is a 'null' return also only ever a 'not found' condition? (I know a profile doc should be public, so unauthorized shouldn't happen, but I assume a generic server error (i.e., a HTTP 500) might be possible, in which case being told the profile doc is 'not found' might be confusing. So perhaps this is a more general error-handling question (i.e., how does the caller of `getSolidDataset()` get helpful error information), but for now, perhaps the message here could be more generic, and be 'Profile document could not be retrieved' (as opposed to 'not found')...?

Reply:
55884346